src/Workspaces/Remote/ServiceHub/Services/KeepAlive/RemoteKeepAliveService.cs

```diff
@@ -15,23 +17,65 @@ protected override IRemoteKeepAliveService CreateService(in ServiceConstructionA
+    public async ValueTask KeepAliveAsync(
         Checksum solutionChecksum,
+        long sessionId,
         CancellationToken cancellationToken)
     {
-        // First get the solution, ensuring that it is currently pinned.
-        return RunServiceAsync(solutionChecksum, async solution =>
+        // Ensure we have a completion source for this sessionId.  Note: we are potentially racing with
+        // WaitForSessionIdAsync on the host side, so it's possible that it will beat us to creating the entry in this
+        // dictionary.  That's fine though, as both calls will be guaranteed to get the same completionSource thanks to
+        // GetOrAdd.
+        var completionSource = GetSessionCompletionSource(sessionId);
+        try
+        {
+
+            // First get the solution, ensuring that it is currently pinned.
+            await RunServiceAsync(solutionChecksum, async solution =>
+            {
+                // Now that we have the solution, we can trigger the completion source to let the host know it can 
+                // proceed with its work.
+                completionSource.TrySetResult(true);
+
+                // Wait for our caller to tell us to cancel.  That way we can release this solution and allow it
+                // to be collected if not needed anymore.
+                //
+                // This was provided by stoub as an idiomatic way to wait indefinitely until a cancellation token triggers.
+                await Task.Delay(-1, cancellationToken).ConfigureAwait(false);
+            }, cancellationToken).ConfigureAwait(false);
+        }
+        finally
         {
-            // Wait for our caller to tell us to cancel.  That way we can release this solution and allow it
-            // to be collected if not needed anymore.
-            //
-            // This was provided by stoub as an idiomatic way to wait indefinitely until a cancellation token triggers.
-            await Task.Delay(-1, cancellationToken).ConfigureAwait(false);
-        }, cancellationToken);
+            // Now that we've been cancelled, remove the completion source from our map.  Note: this cancellationToken
+            // is the one owned by RemoteKeepAliveSession, and only triggered in its Dispose method.  We only get to
+            // that point once the RemoteKeepAliveSession is fully created, returned back out to the caller, and then
+            // eventually Disposed of.  That means that by the time we get here, WaitForSessionIdAsync already must have
+            // returned, so there is no concern about racing at this point, and we are certain to clear out the value
+            // and not leave around stale entries.
+            Contract.ThrowIfFalse(_sessionIdToCompletionSource.TryRemove(sessionId, out var foundSource));
+            Contract.ThrowIfFalse(foundSource == completionSource);
+        }
+    }
+
+    public async ValueTask WaitForSessionIdAsync(long sessionId, CancellationToken cancellationToken)
```

Comment:
What is the scenario where you'd call KeepAliveAsync, but not WaitForSessionIdAsync (could they be combined into one)?

Wouldn't any keep alive session need to wait for the solution to be hydrated?